integration-tests/src/test/java/com/datastax/oss/driver/core/session/ListenersIT.java

```diff
@@ -90,71 +94,102 @@ public void should_inject_session_in_listeners() throws Exception {
                         .build())
                 .build()) {
 
+      // These callbacks are invoked from different threads, ones called from adminExecutor may not
+      // have been called before session initialization has completed so we may have to wait a
+      // little.
+      // Session init thread:
+      // - NodeStateListener#onSessionReady
+      // - RequestTracker#onSessionReady
+      // AdminExecutor thread:
+      // - NodeStateListener#onUp
+      // - SchemaChangeListener#onSessionReady
+      // - *#close
       InOrder inOrder1 = inOrder(nodeListener1);
-      inOrder1.verify(nodeListener1).onSessionReady(session);
-      inOrder1.verify(nodeListener1).onUp(nodeCaptor1.capture());
+      inOrder1.verify(nodeListener1, timeout(500).times(1)).onSessionReady(session);
+      inOrder1.verify(nodeListener1, timeout(500).times(1)).onUp(nodeCaptor1.capture());
 
       InOrder inOrder2 = inOrder(nodeListener2);
-      inOrder2.verify(nodeListener2).onSessionReady(session);
-      inOrder2.verify(nodeListener2).onUp(nodeCaptor2.capture());
+      inOrder2.verify(nodeListener2, timeout(500).times(1)).onSessionReady(session);
+      inOrder2.verify(nodeListener2, timeout(500).times(1)).onUp(nodeCaptor2.capture());
 
       assertThat(nodeCaptor1.getValue().getEndPoint())
           .isEqualTo(SIMULACRON_RULE.getContactPoints().iterator().next());
 
       assertThat(nodeCaptor2.getValue().getEndPoint())
           .isEqualTo(SIMULACRON_RULE.getContactPoints().iterator().next());
 
-      verify(schemaListener1).onSessionReady(session);
-      verify(schemaListener2).onSessionReady(session);
-
-      verify(requestTracker1).onSessionReady(session);
-      verify(requestTracker2).onSessionReady(session);
-
-      assertThat(MyNodeStateListener.onSessionReadyCalled).isTrue();
-      assertThat(MyNodeStateListener.onUpCalled).isTrue();
-
-      assertThat(MySchemaChangeListener.onSessionReadyCalled).isTrue();
-
-      assertThat(MyRequestTracker.onSessionReadyCalled).isTrue();
+      verify(schemaListener1, timeout(500).times(1)).onSessionReady(session);
```

Comment:
This seems like the heart of the problem; an event that's called from a thread totally unrelated to the thread governing where we're at in this code.  Seems like this is the truly essential change we need here.

Reply:
Thank you for the review! You are correct, only the schema listener's callbacks are not being waited for by other mechanisms (session init/close) so these should be the only ones that need to change. I think we also need to keep the code that waits for MySchemaChangeListener#onSessionReady because the bg thread may not have gone through all the schema listeners, even if we verfied one was called earlier. I'll revert all the unnecessary synchronization/waiting and update the comments with my learnings here.